### PR TITLE
peer: calculate replicated index based only on voters

### PR DESCRIPTION
**Pull request by overvenus**
## What have you changed? (mandatory)

Only consider voters when calculating `replicated_idx`, because
learners may be down for a long time. We don't want to pile up
raft logs, and too many logs may cause serious write amplification.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

### CODE REVIEW

**Comment by BusyJay** [comment]
But frequent snapshot can also slow down leader. I don't think there are much differences from learners and voters.

**Comment by siddontang** [comment]
@overvenus 

What is the problem do you meet? using this way can help you fix it? If yes, please show the result here to prove it.

@BusyJay if you think this can also hurt the performance, we should also give the evidence.  Or do we have another way to work around it?

**Comment by BusyJay** [comment]
> if you think this can also hurt the performance, we should also give the evidence

It has been reported before that too many snapshots can consume a lot of IO.

>  Or do we have another way to work around it?

I think it's a common problem not just limited to learner. For example, a crash follower can also lead to a lot of pending logs. So to solve the problem, we need to detect crash nodes easily. For example, using timeout instead of log gap count to gc raft logs or using connection event to verify whether a node is unable to be connected.

File: src/raftstore/store/fsm/peer.rs [hunk 1]
```diff
@@ -2385,10 +2385,13 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
         // `alive_cache_idx` is the smallest `replicated_index` of healthy up nodes.
         // `alive_cache_idx` is only used to gc cache.
+        // Only consider voters when calculating `replicated_idx`, because
+        // learners may be down for a long time. We don't want to pile up
+        // raft logs, and too many logs may cause serious write amplification.
         let truncated_idx = self.fsm.peer.get_store().truncated_index();
         let last_idx = self.fsm.peer.get_store().last_index();
         let (mut replicated_idx, mut alive_cache_idx) = (last_idx, last_idx);
-        for (peer_id, p) in self.fsm.peer.raft_group.raft.prs().iter() {
+        for (peer_id, p) in self.fsm.peer.raft_group.raft.prs().voters() {
```
Comment (hicqu) [inline src/raftstore/store/fsm/peer.rs]: How about also check `pending_request_snapshot` field of the learners?

**Pull request closed without merging**